lib/pbench/server/api/resources/upload_api.py

```diff
@@ -429,32 +415,24 @@ def _put(self, args: ApiParams, request: Request, context: ApiContext) -> Respon
             # the switch to object store. For now we ignore it.
             recovery.add(tarball.delete)
 
-            # Now that we have the tarball, extract the dataset timestamp from
-            # the metadata.log file.
-            #
-            # If the metadata.log is missing or corrupt, or doesn't contain the
-            # "date" property in the "pbench" section, the resulting exception
-            # will cause the upload to fail with an error.
-            #
-            # NOTE: The full metadata.log (as a JSON object with section names
-            # as the top level key) will be stored as a Metadata key using the
-            # reserved internal key "metalog". For retrieval, the "dataset" key
-            # provides a JSON mapping of the Dataset SQL object, enhanced with
-            # the dataset's "metalog" Metadata key value.
+            # Add the processed tarball metadata.log file contents, if any.
             #
-            # NOTE: we're setting the Dataset "created" timestamp here, but it
-            # won't be committed to the database until the "advance" operation
-            # at the end.
+            # If we were unable to find or parse the tarball's metadata.log
+            # file, construct a minimal metadata context identifying the
+            # dataset as a "foreign" benchmark script, and disable indexing,
+            # which requires the metadata.
             try:
+                current_app.logger.info("Processing metadata: {}", tarball.metadata)
                 metalog = tarball.metadata
-                dataset.created = UtcTimeHelper.from_string(
-                    metalog["pbench"]["date"]
-                ).utc_time
+                if not metalog:
+                    metalog = {"pbench": {"script": "Foreign"}}
+                    metadata[Metadata.SERVER_ARCHIVE] = True
+                    attributes["missing_metadata"] = True
                 Metadata.create(dataset=dataset, key=Metadata.METALOG, value=metalog)
```

Comment:
Wrong context. *This* deals with the `metadata.log` values, which become the `dataset.metalog` metadata key. The `metadata` dict defines general metadata across any namespace that the caller can define. That's handled separately, below.

Reply:
So, if the user tries to specify values for `metadata.log` keys in the request, those will be (silently?) ignored?

Reply:
Right now, `metadata.log` (and therefore the `dataset.metalog` metadata) captures agent context at benchmark postprocessing and collection time: it can't be changed later. If someone specifies `?metadata=dataset.metalog.pbench.script:foo` on the upload, it'll fail the check against `Metadata.USER_UPDATEABLE_METADATA` long before we get here, and that's recorded in the `errors` output. (See the upload test case `test_upload_invalid_metadata`.) So they'll get back a 400 and something like `{"message": "at least one specified metadata key is invalid", "errors": ["Key dataset.metalog.pbench.script is invalid or isn't settable"]}`